### PR TITLE
feat: default value calculations run on existing entities

### PR DESCRIPTION
**Pull request by tskimmett**
When initializing an "existing" entity, default value calculation rules will run for properties with no value specified in the initial state for the entity.

```ts
// assuming Name is defaulted to "John Doe"
const e1 = await SomeType.create({ Id: "1" });
console.log(e1.Name);  // John Doe

const e2 = await SomeType.create({ Id: "2", Name: null });
console.log(e2.Name);  // null
```

### CODE REVIEW

File: src/property.ts [hunk 1]
```diff
@@ -840,73 +840,6 @@ export function Property$generateOwnProperty(property: Property, obj: Entity): v
 	});
 }
 
-// TODO: Get rid of `Property$_generateOwnPropertyWithClosure`...
```
Comment (tskimmett) [inline src/property.ts:843]: I did the TODO!

File: src/property.ts [hunk 2]
```diff
@@ -984,16 +917,16 @@ export function Property$init(property: Property, obj: Entity, val: any): void {
...
-		Property$pendingInit(obj, property, true);
-
 		// Do not initialize calculated properties. Calculated properties should be initialized using a property get rule.
 		if (!property.isCalculated) {
 			Property$init(property, obj, Property$getInitialValue(property));
 		}
+
+		// Mark the property as pending initialization if it still has no underlying value to allow default calculation rules to run for it
+		if (obj.__fields__[property.name] === null)
+			Property$pendingInit(obj, property, true);
```
Comment (tskimmett) [inline src/property.ts:929]: I adapted this from exoweb:
image

File: src/property.ts [hunk 3]
```diff
@@ -984,15 +917,17 @@ export function Property$init(property: Property, obj: Entity, val: any): void {
...
-		// Mark the property as pending initialization
 		Property$pendingInit(obj, property, true);
 
 		// Do not initialize calculated properties. Calculated properties should be initialized using a property get rule.
 		if (!property.isCalculated) {
 			Property$init(property, obj, Property$getInitialValue(property));
+
+			// Mark the property as pending initialization if it still has no underlying value to allow default calculation rules to run for it
+			if (obj.__fields__[property.name] === null)
+				Property$pendingInit(obj, property, true);
```
Comment (tskimmett) [inline src/property.ts:930]: I adapted this from exoweb:
image

File: src/property.ts (hunk 1 above)
Comment (mattheyan) [reply src/property.ts:843]: 🤣

File: src/property.ts (hunk 3 above)
Comment (mattheyan) [reply src/property.ts:930]: Oh, now I see that you changed the `isCalculated` logic to not set it when its a default value calculation 😁

File: src/calculated-property-rule.ts [hunk 4]
```diff
@@ -32,14 +32,13 @@ export class CalculatedPropertyRule extends Rule {
...
-					options.returns = [property];
-				}
-				else {
+				if (options.isDefaultValue) {
 					// Ensure the default value rule runs on init of a new instance
 					(options as RuleInvocationOptions).onInitNew = true;
 				}
+
+				// indicate that the rule is responsible for returning the value of the calculated property
+				options.returns = [property];
```
Comment (mattheyan) [inline src/calculated-property-rule.ts:41]: The way I'm reading this, default value calculations would have never run for existing entities? Or, at least not unless one of their inputs changed. So, with this change, they will also run on first access if needed.

File: src/calculated-property-rule.ts [hunk 5]
```diff
@@ -68,8 +67,8 @@ export class CalculatedPropertyRule extends Rule {
 		// register the rule with the target property
 		this.property.rules.push(this);
 
-		// mark the property as calculated if the rule runs on property access
-		if (this.invocationTypes & RuleInvocationType.PropertyGet)
+		// mark the property as calculated if the rule runs on property access and is not a default value calculation
+		if (!options.isDefaultValue && this.invocationTypes & RuleInvocationType.PropertyGet)
```
Comment (mattheyan) [inline src/calculated-property-rule.ts:71]: Why was this change needed? I'm guessing previously there was some reason why the default value calculation would not have run on first access. Was it just because the `returns` option wasn't set, or was there another reason as well?

Comment (mattheyan) [reply src/calculated-property-rule.ts:71]: Is it because the serialization/deserialization code ignores properties with `isCalculated` = true?

**Review by mattheyan — APPROVED** [review]
LGTM

File: src/calculated-property-rule.ts (hunk 5 above)
Comment (tskimmett) [reply src/calculated-property-rule.ts:71]: Yes, if we didn't unset `isCalculated`, the default calculation rule would always overwrite the value. For example, when you load an existing entry that has a value other than what the default would be.

File: src/calculated-property-rule.ts (hunk 4 above)
Comment (tskimmett) [reply src/calculated-property-rule.ts:41]: You're correct. Previously, default calculations would only run when a dependency changed for an existing entity.
